src/legacy/core_plugins/kibana/public/discover/np_ready/angular/discover.js

```diff
@@ -188,23 +177,115 @@ function discoverController(
   $scope,
   $timeout,
   $window,
-  AppState,
   Promise,
   config,
   kbnUrl,
   localStorage,
-  uiCapabilities,
-  getAppState,
-  globalState
+  uiCapabilities
 ) {
-  const filterStateManager = new FilterStateManager(globalState, getAppState, filterManager);
+  const subscriptions = new Subscription();
+  const $fetchObservable = new Subject();
+  let inspectorRequest;
+  const savedSearch = $route.current.locals.savedObjects.savedSearch;
+  $scope.searchSource = savedSearch.searchSource;
+  $scope.indexPattern = resolveIndexPatternLoading();
+  $scope.timeRangeObj = timefilter.getTime();
+  $scope.refreshInterval = timefilter.getRefreshInterval();
+
+  const getTimeField = () => {
+    return indexPatternsUtils.isDefault($scope.indexPattern)
+      ? $scope.indexPattern.timeFieldName
+      : undefined;
+  };
+
+  const {
+    appStateContainer,
+    startSync: startStateSync,
+    stopSync: stopStateSync,
+    setAppState,
+    replaceUrlAppState,
+    isAppStateDirty,
+    kbnUrlStateStorage,
+    getPreviousAppState,
+  } = getState({
+    defaultAppState: getStateDefaults(),
+    storeInSessionStorage: config.get('state:storeInSessionStorage'),
+  });
+  if (appStateContainer.getState().index !== $scope.indexPattern.id) {
+    //used index pattern is different than the given by url/state which is invalid
+    setAppState({ index: $scope.indexPattern.id });
+  }
+  $scope.state = { ...appStateContainer.getState() };
+
+  // syncs `_g` portion of url with query services
+  const { stop: stopSyncingGlobalStateWithUrl } = syncQueryStateWithUrl(
+    data.query,
+    kbnUrlStateStorage
+  );
+
+  const stopSyncingQueryAppStateWithStateContainer = connectToQueryState(
```

Comment:
Noticed an issue that app filters are removed on page reload. 
Basically in the beginning, whatever is in filter manager takes precedences over app filters in url. 
This also causes side effect - when navigating from dashboard with app filters to discover, dashboard's app filters are preserved. (because filterManger state is preserved during navigation). 

I guess this is the line which prevents this behaviour in dashboard: https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/kibana/public/dashboard/np_ready/dashboard_app_controller.tsx#L152

Reply:
good catch! thx! also fixes the 'New' issue you reported